src/core/border_router/routing_manager.cpp

```diff
@@ -578,11 +578,25 @@ void RoutingManager::EvaluateOnLinkPrefix(void)
             LogInfo("Start advertising on-link prefix %s on %s", mLocalOnLinkPrefix.ToString().AsCString(),
                     mInfraIf.ToString().AsCString());
 
-            // Call `InvalidateDiscoveredPrefixes()` after setting
-            // `mIsAdvertisingLocalOnLinkPrefix` to remove on-link
-            // prefix in case it was discovered and included in
-            // `mDiscoveredPrefixes` list earlier.
-            InvalidateDiscoveredPrefixes();
+            // We go through `mDiscoveredPrefixes` list to check if the
```

Comment:
The original code has the advantage of removing invalid prefixes in one dedicated method, and sets up `mDiscoveredPrefixInvalidTimer` correctly, which I think is even smaller in code size.

Do we fail to reset the fire time for `mDiscoveredPrefixInvalidTimer`?

Reply:
Yes. It is correct that we don't re-schedule the timer in this case but I think it is fine. If the timer happens to be scheduled for the entry that we removed, once it fires we go through and re-schedule it (if needed). 
- In general, I actually slightly favor if we avoid canceling/updating a running timer if we don't need to. We can have a lot of running `Timer` objects in OT stack (they are are added in a linked-list) and removing or re-scheduling one can take more CPU time compared to letting it fire (and then re-scheduling it if we need to when it fires).
- And in some cases it can lead to smaller code size (to be fair, the case here may not be a good example of this but in other situations not canceling a timer may result in smaller code size :)...)

-------

The background for this is that I want to later add a class for `DiscoveredPrefixes` (to maintain the list of prefixes from processed RA message)
- It can have its own timer and age out the prefixes, and provide simple methods to `RoutingManager`:
   -  e.g. methods to add prefix to list, remove a prefix and unpublish it, remove prefix without unpublishing, mechansim to signal back when it gets empty or has no on-link prefix, etc.
   - It can represents itself as if it is two different lists, a list of PrefixInfo options (discovered on-link prefixes), and a list of RouteInfo options (which we can use for OMR or later for default route) but then internally use the same underlying `Array` to store all entries.
- This PR is a step towards this (trying to decouple the logic).
- I feel adding a class for `DiscoveredPrefixes` may help simplify the code. Any thoughts/suggestions?

Reply:
> Yes. It is correct that we don't re-schedule the timer in this case but I think it is fine. If the timer happens to be scheduled for the entry that we removed, once it fires we go through and re-schedule it (if needed).
> 
> * In general, I actually slightly favor if we avoid canceling/updating a running timer if we don't need to. We can have a lot of running `Timer` objects in OT stack (they are are added in a linked-list) and removing or re-scheduling one can take more CPU time compared to letting it fire (and then re-scheduling it if we need to when it fires).

Should be fine since `InvalidateDiscoveredPrefixes` only reschedule the timer once? 

> * And in some cases it can lead to smaller code size (to be fair, the case here may not be a good example of this but in other situations not canceling a timer may result in smaller code size :)...)

The code size point I made is that just calling `InvalidateDiscoveredPrefixes()` is smaller in code size than the new code here. 

I guess code size is usually not a bottleneck for BRs. 

> The background for this is that I want to later add a class for `DiscoveredPrefixes` (to maintain the list of prefixes from processed RA message)
> 
> * It can have its own timer and age out the prefixes, and provide simple methods to `RoutingManager`:
>   
>   * e.g. methods to add prefix to list, remove a prefix and unpublish it, remove prefix without unpublishing, mechansim to signal back when it gets empty or has no on-link prefix, etc.
>   * It can represent itself as if it is two different lists, a list of PrefixInfo options (discovered on-link prefixes), and a list of RouteInfo options (which we can use for OMR or later for default route) but then internally use the same underlying `Array` to store all entries.
> * This PR is a step towards this (trying to decouple the logic).
> * I feel adding a class for `DiscoveredPrefixes` may help simplify the code. Any thoughts/suggestions?

I like the idea of `DiscoveredPrefixes`. 

---

This PR LGTM, considering it's an intermediate state.